File: backend/app/routers/web_session_profiles.py
from __future__ import annotations

from fastapi import APIRouter, Depends, status

from backend.app.config import AppConfig
from backend.app.dependencies import get_config, get_web_session_profile_store
from backend.app.schemas.web_session_profiles import (
    WebSessionProfileCreateRequest,
    WebSessionProfileDeleteResponse,
    WebSessionProfileEnvelope,
    WebSessionProfileListEnvelope,
    WebSessionProfileLoginRequest,
    WebSessionProfileResponse,
    WebSessionProfileUpdateRequest,
)
from backend.app.services.web_session_profiles import (
    WebSessionProfileService,
    WebSessionProfileStore,
    run_managed_session_login,
)


router = APIRouter(prefix="/web/session-profiles", tags=["web-session-profiles"])


@router.get("", response_model=WebSessionProfileListEnvelope)
def list_web_session_profiles(
    config: AppConfig = Depends(get_config),
    store: WebSessionProfileStore = Depends(get_web_session_profile_store),
) -> WebSessionProfileListEnvelope:
    service = WebSessionProfileService(store=store, app_data_dir=config.app_data_dir)
    return WebSessionProfileListEnvelope(
        profiles=[WebSessionProfileResponse(**item) for item in service.list_profiles()]
    )


@router.post("", response_model=WebSessionProfileEnvelope, status_code=status.HTTP_201_CREATED)
def create_web_session_profile(
    payload: WebSessionProfileCreateRequest,
    config: AppConfig = Depends(get_config),
    store: WebSessionProfileStore = Depends(get_web_session_profile_store),
) -> WebSessionProfileEnvelope:
    service = WebSessionProfileService(store=store, app_data_dir=config.app_data_dir)
    return WebSessionProfileEnvelope(
        profile=WebSessionProfileResponse(
            **service.create_profile(
                name=payload.name,
                mode=payload.mode,
                browser_channel=payload.browser_channel,
                profile_path=payload.profile_path,
                login_url=payload.login_url,
            )
        )
    )


@router.patch("/{profile_id}", response_model=WebSessionProfileEnvelope)
def update_web_session_profile(
    profile_id: str,
    payload: WebSessionProfileUpdateRequest,
    config: AppConfig = Depends(get_config),
    store: WebSessionProfileStore = Depends(get_web_session_profile_store),
) -> WebSessionProfileEnvelope:
    service = WebSessionProfileService(store=store, app_data_dir=config.app_data_dir)
    return WebSessionProfileEnvelope(
        profile=WebSessionProfileResponse(
            **service.update_profile(
                profile_id,
                name=payload.name,
                browser_channel=payload.browser_channel,
                profile_path=payload.profile_path,
                login_url=payload.login_url,
            )
        )
    )


@router.post("/{profile_id}/login", response_model=WebSessionProfileEnvelope, status_code=status.HTTP_202_ACCEPTED)
def start_web_session_profile_login(
    profile_id: str,
    payload: WebSessionProfileLoginRequest,
    config: AppConfig = Depends(get_config),
    store: WebSessionProfileStore = Depends(get_web_session_profile_store),
) -> WebSessionProfileEnvelope:
    service = WebSessionProfileService(
        store=store,
        app_data_dir=config.app_data_dir,
        login_runner=run_managed_session_login,
    )
    return WebSessionProfileEnvelope(
        profile=WebSessionProfileResponse(
            **service.start_managed_login(profile_id, login_url=payload.login_url)
        )
    )


@router.delete("/{profile_id}", response_model=WebSessionProfileDeleteResponse)
def delete_web_session_profile(
    profile_id: str,
    config: AppConfig = Depends(get_config),
    store: WebSessionProfileStore = Depends(get_web_session_profile_store),
) -> WebSessionProfileDeleteResponse:
    service = WebSessionProfileService(store=store, app_data_dir=config.app_data_dir)
    return WebSessionProfileDeleteResponse(deleted=service.delete_profile(profile_id))

File: apps/desktop/src/renderer/stores/webSessionProfiles.ts
import { computed, ref } from "vue";
import { defineStore } from "pinia";

import { NetworkError, api } from "@/services/api";
import type { WebSessionProfile, WebSessionProfileCreateRequest } from "@/types";

export const useWebSessionProfilesStore = defineStore("webSessionProfiles", () => {
  const profiles = ref<WebSessionProfile[]>([]);
  const loading = ref(false);
  const saving = ref(false);
  const loggingInProfileId = ref<string | null>(null);
  const deletingProfileId = ref<string | null>(null);
  const error = ref<string | null>(null);
  const connectionUnavailable = ref(false);

  const profileOptions = computed(() =>
    profiles.value.map((profile) => ({
      label: `${profile.name} · ${profile.mode === "browser_profile" ? "浏览器" : "应用内"}`,
      value: profile.id
    }))
  );

  const loadProfiles = async () => {
    loading.value = true;
    error.value = null;
    try {
      const response = await api.listWebSessionProfiles();
      profiles.value = response.profiles;
      connectionUnavailable.value = false;
      return response.profiles;
    } catch (errorValue) {
      connectionUnavailable.value = errorValue instanceof NetworkError;
      error.value = (errorValue as Error).message;
      throw errorValue;
    } finally {
      loading.value = false;
    }
  };

  const createProfile = async (payload: WebSessionProfileCreateRequest) => {
    saving.value = true;
    error.value = null;
    try {
      const response = await api.createWebSessionProfile(payload);
      profiles.value = [response.profile, ...profiles.value.filter((item) => item.id !== response.profile.id)];
      connectionUnavailable.value = false;
      return response.profile;
    } catch (errorValue) {
      connectionUnavailable.value = errorValue instanceof NetworkError;
      error.value = (errorValue as Error).message;
      throw errorValue;
    } finally {
      saving.value = false;
    }
  };

  const startManagedLogin = async (profileId: string, loginUrl?: string | null) => {
    loggingInProfileId.value = profileId;
    error.value = null;
    try {
      const response = await api.startWebSessionProfileLogin(profileId, {
        login_url: loginUrl ?? null
      });
      profiles.value = profiles.value.map((item) => (item.id === profileId ? response.profile : item));
      connectionUnavailable.value = false;
      return response.profile;
    } catch (errorValue) {
      connectionUnavailable.value = errorValue instanceof NetworkError;
      error.value = (errorValue as Error).message;
      throw errorValue;
    } finally {
      loggingInProfileId.value = null;
    }
  };

  const deleteProfile = async (profileId: string) => {
    deletingProfileId.value = profileId;
    error.value = null;
    try {
      await api.deleteWebSessionProfile(profileId);
      profiles.value = profiles.value.filter((item) => item.id !== profileId);
      connectionUnavailable.value = false;
      return true;
    } catch (errorValue) {
      connectionUnavailable.value = errorValue instanceof NetworkError;
      error.value = (errorValue as Error).message;
      throw errorValue;
    } finally {
      deletingProfileId.value = null;
    }
  };

  return {
    profiles,
    loading,
    saving,
    loggingInProfileId,
    deletingProfileId,
    error,
    connectionUnavailable,
    profileOptions,
    loadProfiles,
    createProfile,
    startManagedLogin,
    deleteProfile
  };
});
